File: sdk/js/packages/next/src/server/admin.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import { createAdminClient } from './admin.js';

const config = { baseURL: 'http://localhost:8080', apiKey: 'bnc_api_test' };

describe('bouncingAdmin', () => {
  afterEach(() => vi.restoreAllMocks());

  it('listUsers sends correct request', async () => {
    const mockFetch = vi.fn(async () => ({
      ok: true,
      json: async () => ({ users: [], total: 0 }),
    } as Response));
    vi.stubGlobal('fetch', mockFetch);

    const admin = createAdminClient(config);
    await admin.listUsers({ page: 2, perPage: 10, q: 'alice' });

    const [url, init] = mockFetch.mock.calls[0] as unknown as [string, RequestInit];
    expect(url).toContain('/manage/users');
    expect(url).toContain('page=2');
    expect(url).toContain('per_page=10');
    expect(url).toContain('q=alice');
    expect((init.headers as Record<string, string>)['Authorization']).toBe('Bearer bnc_api_test');
  });

  it('inviteUser sends correct body', async () => {
    const mockFetch = vi.fn(async () => ({
      ok: true,
      json: async () => ({ id: 'u1', email: 'a@b.com', status: 'pending' }),
    } as Response));
    vi.stubGlobal('fetch', mockFetch);

    const admin = createAdminClient(config);
    const result = await admin.inviteUser({ email: 'a@b.com', role: 'editor' });

    const [url, init] = mockFetch.mock.calls[0] as unknown as [string, RequestInit];
    expect(url).toContain('/manage/users/invite');
    expect(init.method).toBe('POST');
    expect(JSON.parse(init.body as string)).toMatchObject({ email: 'a@b.com', role: 'editor' });
    expect(result.status).toBe('pending');
  });

  it('deleteUser uses DELETE method', async () => {
    const mockFetch = vi.fn(async () => ({
      ok: true,
      json: async () => ({ deleted: true }),
    } as Response));
    vi.stubGlobal('fetch', mockFetch);

    const admin = createAdminClient(config);
    await admin.deleteUser('user-abc');

    const [url, init] = mockFetch.mock.calls[0] as unknown as [string, RequestInit];
    expect(url).toContain('/manage/users/user-abc');
    expect(init.method).toBe('DELETE');
  });

  it('throws on API error response', async () => {
    vi.stubGlobal('fetch', vi.fn(async () => ({
      ok: false,
      status: 404,
      json: async () => ({ error: { message: 'user not found' } }),
    } as Response)));

    const admin = createAdminClient(config);
    await expect(admin.deleteUser('missing')).rejects.toThrow('user not found');
  });
});

File: sdk/go/admin_test.go
package bouncing

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestAdminListUsers(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/manage/users" {
			t.Errorf("path: got %q", r.URL.Path)
		}
		if r.Header.Get("Authorization") != "Bearer test-key" {
			t.Errorf("auth: got %q", r.Header.Get("Authorization"))
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"users": []map[string]any{
				{"id": "u1", "email": "a@b.com", "status": "active"},
			},
			"total": 1,
		})
	}))
	defer srv.Close()

	client := New(Config{BaseURL: srv.URL, APIKey: "test-key"})
	users, err := client.Admin.ListUsers(context.Background())
	if err != nil {
		t.Fatalf("ListUsers: %v", err)
	}
	if len(users) != 1 || users[0].Email != "a@b.com" {
		t.Errorf("users: %+v", users)
	}
}

func TestAdminInviteUser(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("method: got %q", r.Method)
		}
		var body map[string]string
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["email"] != "new@example.com" {
			t.Errorf("email: got %q", body["email"])
		}
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"id": "u2", "email": "new@example.com", "status": "pending",
		})
	}))
	defer srv.Close()

	client := New(Config{BaseURL: srv.URL, APIKey: "test-key"})
	user, err := client.Admin.InviteUser(context.Background(), "new@example.com", "New User", "editor")
	if err != nil {
		t.Fatalf("InviteUser: %v", err)
	}
	if user.Email != "new@example.com" {
		t.Errorf("email: got %q", user.Email)
	}
}

func TestAdminDeleteUser(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "DELETE" || r.URL.Path != "/manage/users/u1" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"deleted": true})
	}))
	defer srv.Close()

	client := New(Config{BaseURL: srv.URL, APIKey: "test-key"})
	err := client.Admin.DeleteUser(context.Background(), "u1")
	if err != nil {
		t.Fatalf("DeleteUser: %v", err)
	}
}

func TestAdminListRoles(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"roles": []map[string]any{
				{"id": "r1", "name": "admin", "permissions": []string{"*"}},
			},
		})
	}))
	defer srv.Close()

	client := New(Config{BaseURL: srv.URL, APIKey: "test-key"})
	roles, err := client.Admin.ListRoles(context.Background())
	if err != nil {
		t.Fatalf("ListRoles: %v", err)
	}
	if len(roles) != 1 || roles[0].Name != "admin" {
		t.Errorf("roles: %+v", roles)
	}
}

func TestAdminServerError(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
		_, _ = w.Write([]byte(`{"error":"internal"}`))
	}))
	defer srv.Close()

	client := New(Config{BaseURL: srv.URL, APIKey: "test-key"})
	_, err := client.Admin.ListUsers(context.Background())
	if err == nil {
		t.Error("expected error for 500 response")
	}
}
